README.md
This program scans the Rotoworld website for updates and can email individual
blurbs to a desired email address. It can also send these as 160-character
blocks, so one can send them to a phone as a text message. However, the "email
address" of the phone must be specified in the program (for example, Verizon's
are phonenumber@vtext.com).

As of version 0.7, there is also a rudimentary Swing GUI included with the
program. Executable .jar files of both the CLI and GUI versions of RotoScan
can be found in the [GitHub repository](https://github.com/TheCrittaC/RotoScan).

A screenshot of the GUI version is shown below.

![RotoScan GUI Screenshot](https://i.minus.com/ib2m3J61tI0Ht5.png)

The code to send email via a GMail address was modified from code at [this page](http://www.mkyong.com/java/javamail-api-sending-email-via-gmail-smtp-example/).

This program uses JSoup and JavaMail. Both of these jar files are included in
the GitHub repository if you wish to compile your own version of this software.

JSoup is distributed under the terms of the [MIT License](http://opensource.org/licenses/MIT).

JavaMail is distributed under the terms of the [Oracle Binary Code License
Agreement for Java EE Technologies](http://download.oracle.com/otn-pub/java/licenses/OTN_JavaEE_Legacy_Binary-Code-License_30Jan2012.txt).

My code (all other .java files included in this repository) is licensed under
the GNU General Public license, version two or later. Version two is included
in this GitHub repository as LICENSE.txt.

Changelog.md
# RotoScan Changelog
	## Version 0.7
	* Initial version of the GUI (it's terrible, but it works)	
	## Version 0.5
	* Initial version of program
	* Rudimentary CLI scans all sections of Rotoworld for new blurbs and
	emails them or sends them as text messages (emails in 160-character
	chunks)
	* Checks for new blurbs to desired sections of Rotoworld once a minute


RotoScan.java
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.*;
import java.util.Scanner;
import java.lang.Thread;
public class RotoScan{
    public static void main(String args[]){
	Scanner scan = new Scanner(System.in);
	String input;
	String email = "";
	String text = "";
	String to = "";
	String password = ""; //initial strings that are initialized
	String NFLURL = "http://www.rotoworld.com/playernews/nfl/football/";
	String MLBURL = "http://www.rotoworld.com/playernews/mlb/baseball/";
	String NBAURL = "http://www.rotoworld.com/playernews/nba/basketball/";
	String NHLURL = "http://www.rotoworld.com/playernews/nhl/hockey/";
	String PGAURL = "http://www.rotoworld.com/playernews/gol/golf/";
	String NASCARURL = "http://www.rotoworld.com/playernews/nas/NASCAR/";
	String CFBURL = "http://www.rotoworld.com/playernews/cfb/nfl-draft"; //URLs
	Thread NFLThread;
	Thread MLBThread;
	Thread NBAThread;
	Thread NHLThread;
	Thread PGAThread;
	Thread NASCARThread;
	Thread CFBThread; //threads
	ScanRoto NFLRoto;
	ScanRoto MLBRoto;
	ScanRoto NBARoto;
	ScanRoto NHLRoto;
	ScanRoto PGARoto;
	ScanRoto NASCARRoto;
	ScanRoto CFBRoto; //ScanRoto objects

	System.out.println("Do you want this program to send emails?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    System.out.println("What is your email address? GMail only at this time.");
	    email = scan.nextLine();
	    System.out.println("Where will these emails be going?");
	    to = scan.nextLine();
	    System.out.println("What is your GMail password?");
	    password = scan.nextLine();
	} //gets user email information
	System.out.println("Do you want to send text messages?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    if ((email.equals(""))){
		System.out.println("What is your email address? GMail only at this time.");
		email = scan.nextLine();
		System.out.println("What is your GMail password?");
		password = scan.nextLine();
	    } //gets email account info if it is not already entered
	    System.out.println("Where will these messages be going? Enter an email address.");
	    text = scan.nextLine();
	}
	System.out.println("Do you want to scan for football updates?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    NFLRoto = new ScanRoto(NFLURL, email, to, password, text);
	    NFLThread = new Thread(NFLRoto, "NFL");
	    NFLThread.start();
	} //creates NFL thread
	
	System.out.println("Do you want to scan for baseball updates?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    MLBRoto = new ScanRoto (MLBURL, email, to, password, text);
	    MLBThread = new Thread(MLBRoto, "MLB");
	    MLBThread.start();
	} //creates MLB thread
	
	System.out.println("Do you want to scan for basketball updates?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    NBARoto = new ScanRoto (NBAURL, email, to, password, text);
	    NBAThread = new Thread(NBARoto, "NBA");
	    NBAThread.start();
	} //creates NBA thread
	
	System.out.println("Do you want to scan for hockey updates?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    NHLRoto = new ScanRoto (NHLURL, email, to, password, text);
	    NHLThread = new Thread(NHLRoto, "NHL");
	    NHLThread.start();
	} //creates NHL thread
	
	System.out.println("Do you want to scan for Golf updates?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    PGARoto = new ScanRoto (PGAURL, email, to, password, text);
	    PGAThread = new Thread(PGARoto, "PGA");
	    PGAThread.start();
	} //creates PGA thread
	
	System.out.println("Do you want to scan for NASCAR updates?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    NASCARRoto = new ScanRoto (NASCARURL, email, to, password, text);
	    NASCARThread = new Thread(NASCARRoto, "NASCAR");
	    NASCARThread.start();
	} //creates NASCAR thread
	
	System.out.println("Do you want to scan for college football updates?");
	input = scan.nextLine();
	if (input.startsWith("y") || input.startsWith("Y")){
	    CFBRoto = new ScanRoto (CFBURL, email, to, password, text);
	    CFBThread = new Thread(CFBRoto, "CFB");
	    CFBThread.start();
	} //creates college football thread
    }
}

class ScanRoto implements Runnable{
    private String url; //url for the thread
    private static String fromEmail; //source email address
    private static String toEmail; //destination email address
    private static String toText; //destination text messaging address
    private static String password; //GMail password 
    private Document doc; //document for the page
    private Blurb last; //the previous blurb
    private Blurb latest; //the latest blurb
    private Elements blurbs; //all of the blurbs on the page
    private String blurbString; //the string form of the most recent blurb
    public ScanRoto(String url, String fromEmail, String toEmail, String password, String text){
	this.url = url;
	this.fromEmail = fromEmail;
	this.toEmail = toEmail;
	this.password = password;
	this.toText = text; //stores local variables
    }
    public void run(){
	this.last = null; //initially no blurb loaded
	while (true){
	    try{
		this.doc = Jsoup.connect(this.url).get(); //gets the page
	    } catch(Exception e) {
		e.printStackTrace();
	    }
	    this.blurbs = this.doc.select("div[class=pb]"); //creates a list of blurbs
	    this.latest = new Blurb(this.blurbs.first()); //gets first blurb
	    this.blurbString = this.latest.toString(); //gets blurb in string form
	    if (this.last == null || !(this.last.toString().equals(this.latest.toString()))){
		//checks for updates to blurb
		this.last = this.latest; //updates last if there is a new blurb
		if (this.last.isComplete()){
		    //if blurb is complete (has impact), proceed
		    if (!(toEmail.equals("")))
			mailBlurb(blurbString);
		    //email blurb if we have an email address
		    if (!(toText.equals("")))
			textBlurb(blurbString);
		    //texts blurb if we have a text email address
		}
		    
	    }
	    try{
		Thread.sleep(60000); //sleeps for a minute
	    } catch (Exception e) {
		e.printStackTrace();
	    } 
	}
    }
    
    public static void mailBlurb(String text){
	SendEmail.send(fromEmail, toEmail, password, text);
	//sends the email
    }

    public static void textBlurb(String text){
	for (int i = 0; i < text.length(); i+= 159){
	    if (i + 159 > text.length())
		SendEmail.send(fromEmail, toText, password, text.substring(i, text.length()));
	    //sends email as text message in 160-character blocks
	    else
		SendEmail.send(fromEmail, toText, password, text.substring(i, i + 159));
	    //if we are at the end of the string, we send the rest
	}
    }
}


gui/Blurb.java
import org.jsoup.nodes.Element;

public class Blurb{

    private String playerName;
    private String position;
    private String team;
    private String report;
    private String impact;
    private boolean completeBlurb; //local variables


    public Blurb(Element e){
	completeBlurb = true; //assumes blurb is complete unless otherwise set
	this.playerName = e.select("div[class=player]").toString().split(">")[2].split("<")[0];
	this.position = e.select("div[class=player]").toString().split(">")[3].split("- ")[1];
	this.team = e.select("div[class=player]").toString().split(">")[4].split("<")[0];
	this.report = e.select("div[class=report]").toString().replaceAll("<i>", "");
	this.report = this.report.replaceAll("</i>", ""); //remove italics tags
	this.report = this.report.split("<p>")[1].split("</p>")[0];
	this.report = this.report.replaceAll("&quot;", "\"");
	this.report = this.report.replaceAll("\n", ""); //removes newlines
	try{
	    this.impact = e.select("div[class=impact]").toString().replaceAll("<i>", "");
	    this.impact = this.impact.replaceAll("</i>", "");
	    //remove italics tags
	    this.impact = this.impact.split(">\n  ")[1].split("<")[0];
	    //gets the data from the page
	    this.impact = this.impact.replaceAll("&quot;", "\""); 
	    //changes HTML quotes to actual quotes
	    this.impact = this.impact.replaceAll("\n", ""); //removes newlines
	} catch (Exception exc) {
	    this.impact = "";
	    completeBlurb = false;
	    //since there is no impact, the blurb is not complete and we
	    //do not want to send it yet
	}
    }

    public boolean isComplete(){
	return completeBlurb; //returns if blurb is complete or not
    }

    public String toString(){
	return playerName + " - " + position + " - " + team + "\n" + report + "\n" + impact;
	//converts to the following form
	/*
	  Player Name - Position - Team
	  Report
	  Impact */
    }
    

}			  


gui/SendEmail.java
//package com.mkyong.common;
 
import java.util.Properties;
 
import javax.mail.Message;
import javax.mail.MessagingException;
import javax.mail.PasswordAuthentication;
import javax.mail.Session;
import javax.mail.Transport;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
 
public class SendEmail{
 
    public static void send(String from, String to, String password, String text) {
	
	final String username = from;
	final String pass = password;
 
	Properties props = new Properties();
	props.put("mail.smtp.auth", "true");
	props.put("mail.smtp.starttls.enable", "true");
	props.put("mail.smtp.host", "smtp.gmail.com");
	props.put("mail.smtp.port", "587");
 
	Session session = Session.getInstance(props,
					      new javax.mail.Authenticator() {
						  protected PasswordAuthentication getPasswordAuthentication() {
						      return new PasswordAuthentication(username, pass);
						  }
					      });
	try {
 
	    Message message = new MimeMessage(session);
	    message.setFrom(new InternetAddress(username));
	    message.setRecipients(Message.RecipientType.TO,
				  InternetAddress.parse(to));
	    message.setSubject("");
	    message.setText(text);
 
	    Transport.send(message);
 	    
	} catch (MessagingException e) {
	    throw new RuntimeException(e);
	}
    }
}


gui/RotoScanGUI.java
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.*;
import java.util.Scanner;
import java.lang.Thread;
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
public class RotoScanGUI extends Jpanel{

    Scanner scan = new Scanner(System.in);
    String input;
    String email = "";
    String text = "";
    String to = "";
    String password = ""; //initial strings that are initialized
    String NFLURL = "http://www.rotoworld.com/playernews/nfl/football/";
    String MLBURL = "http://www.rotoworld.com/playernews/mlb/baseball/";
    String NBAURL = "http://www.rotoworld.com/playernews/nba/basketball/";
    String NHLURL = "http://www.rotoworld.com/playernews/nhl/hockey/";
    String PGAURL = "http://www.rotoworld.com/playernews/gol/golf/";
    String NASCARURL = "http://www.rotoworld.com/playernews/nas/NASCAR/";
    String CFBURL = "http://www.rotoworld.com/playernews/cfb/nfl-draft"; //URLs

    Thread NFLThread;
    Thread MLBThread;
    Thread NBAThread;
    Thread NHLThread;
    Thread PGAThread;
    Thread NASCARThread;
    Thread CFBThread; //threads

    ScanRoto NFLRoto;
    ScanRoto MLBRoto;
    ScanRoto NBARoto;
    ScanRoto NHLRoto;
    ScanRoto PGARoto;
    ScanRoto NASCARRoto;
    ScanRoto CFBRoto; //ScanRoto objects

    private JCheckBox NFLBox;
    private JCheckBox MLBBox;
    private JCheckBox NBABox;
    private JCheckBox NHLBox;
    private JCheckBox PGABox;
    private JCheckBox NASCARBox;
    private JCheckBox CFBBox; //checkboxes for various sections of Rotoworld

    private JPanel left;
    private JPanel right; //left and right sides of the window

    private JPanel emailBox;
    private JLabel emailLabel;
    private JPanel textBox;
    private JLabel textLabel;
    private JTextField emailField;
    private JTextField textField;
    private JTextField yourEmailField;
    private JLabel yourEmailLabel;
    private JLabel passwordLabel;
    private JPasswordField passwordField;
    private JButton startStop; //right-side components

    void setup(){

	this.setLayout(new FlowLayout());
	left = new JPanel(new GridLayout(7, 1)); //left side is just checkboxes
	NFLBox = new JCheckBox("Football");
	left.add(NFLBox);
	MLBBox = new JCheckBox("Baseball");
	left.add(MLBBox);
	NBABox = new JCheckBox("Basketball");
	left.add(NBABox);
	NHLBox = new JCheckBox("Hockey");
	left.add(NHLBox);
	PGABox = new JCheckBox("Golf");
	left.add(PGABox);
	NASCARBox = new JCheckBox("NASCAR");
	left.add(NASCARBox);
	CFBBox = new JCheckBox("College Football");
	left.add(CFBBox); //creates checkboxes and adds them to left side

	right = new JPanel(new GridLayout(11,1));
	emailBox = new JCheckBox("Send email?");
	right.add(emailBox);
	textBox = new JCheckBox("Send text message?");
	right.add(textBox);
	emailLabel = new JLabel("Destination email address:");
	right.add(emailLabel);
	emailField = new JTextField();
	right.add(emailField);
	textLabel = new JLabel("Destination text message address:");
	right.add(textLabel);
	textField = new JTextField();
	right.add(textField);
	yourEmailLabel = new JLabel("Your email address:");
	right.add(yourEmailLabel);
	yourEmailField = new JTextField();
	right.add(yourEmailField);
	passwordLabel = new JLabel("Your email password:");
	right.add(passwordLabel);
	passwordField = new JPasswordField();
	right.add(passwordField);
	startStop = new JButton("Start!");
	right.add(startStop); //adds right-side components
	
	startStop.addActionListener(new ActionListener(){
		public void actionPerformed(ActionEvent e){
		    if (startStop.getText().equals("Start!")){
			startStop.setText("Stop!");
			if (emailBox.isSelected())
			    email = yourEmailField.getText();
			else
			    email = ""; //sets email
			to = emailField.getText(); //gets destination address
			password = passwordField.getText(); //gets password
			if (textBox.isSelected())
			    text = textField.getText();
			else
			    text = ""; //gets text message address
			if (NFLBox.isSelected()){
			    NFLRoto = new ScanRoto(NFLURL, email, to, password, text);
			    NFLThread = new Thread(NFLRoto, "NFL");
			    NFLThread.start();
			} //creates NFL thread

			if (MLBBox.isSelected()){
			    MLBRoto = new ScanRoto(MLBURL, email, to, password, text);
			    MLBThread = new Thread(MLBRoto, "MLB");
			    MLBThread.start();
			} //creates MLB thread
			    
			if (NBABox.isSelected()){
			    NBARoto = new ScanRoto(NBAURL, email, to, password, text);
			    NBAThread = new Thread(NBARoto, "NBA");
			    NBAThread.start();
			} //creates NBA thread

			if (NHLBox.isSelected()){
			    NHLRoto = new ScanRoto(NHLURL, email, to, password, text);
			    NHLThread = new Thread(NHLRoto, "NHL");
			    NHLThread.start();
			} //creates NHL thread

			if (PGABox.isSelected()){
			    PGARoto = new ScanRoto(PGAURL, email, to, password, text);
			    PGAThread = new Thread(PGARoto, "PGA");
			    PGAThread.start();
			} //creates PGA thread

			if (NASCARBox.isSelected()){
			    NASCARRoto = new ScanRoto(NASCARURL, email, to, password, text);
			    NASCARThread = new Thread(NASCARRoto, "NASCAR");
			    NASCARThread.start();
			} //creates NASCAR thread

			if (CFBBox.isSelected()){
			    CFBRoto = new ScanRoto(CFBURL, email, to, password, text);
			    CFBThread = new Thread(CFBRoto, "CFB");
			    CFBThread.start();
			} //creates college football thread
			NFLBox.setEnabled(false);
			MLBBox.setEnabled(false);
			NBABox.setEnabled(false);
			NHLBox.setEnabled(false);
			PGABox.setEnabled(false);
			NASCARBox.setEnabled(false);
			CFBBox.setEnabled(false); //disables all checkboxes
			//once the start button is clicked
		    }

		    else{
			if (NFLBox.isSelected())
			    NFLRoto.stop();
			if (MLBBox.isSelected())
			    MLBRoto.stop();
			if (NBABox.isSelected())
			    NBARoto.stop();
			if (NHLBox.isSelected())
			    NHLRoto.stop();
			if (PGABox.isSelected())
			    PGARoto.stop();
			if (NASCARBox.isSelected())
			    NASCARRoto.stop();
			if (CFBBox.isSelected())
			    CFBRoto.stop(); //stops all running threads
			startStop.setText("Start!");
			NFLBox.setEnabled(true);
			MLBBox.setEnabled(true);
			NBABox.setEnabled(true);
			NHLBox.setEnabled(true);
			PGABox.setEnabled(true);
			NASCARBox.setEnabled(true);
			CFBBox.setEnabled(true); //enables all checkboxes
		    }
		});
	    }
	
	
    }
    public RotoScanGUI(){
	setup(); //we create the panel
    }

    
	
}

    class ScanRoto implements Runnable{
	private String url; //url for the thread
	private static String fromEmail; //source email address
	private static String toEmail; //destination email address
	private static String toText; //destination text messaging address
	private static String password; //GMail password 
	private Document doc; //document for the page
	private Blurb last; //the previous blurb
	private Blurb latest; //the latest blurb
	private Elements blurbs; //all of the blurbs on the page
	private String blurbString; //the string form of the most recent blurb
	private boolean running;
	public ScanRoto(String url, String fromEmail, String toEmail, String password, String text){
	    this.url = url;
	    this.fromEmail = fromEmail;
	    this.toEmail = toEmail;
	    this.password = password;
	    this.toText = text; //stores local variables
	    this.running = true; //sets thread state to running
	}

	public void stop(){
	    running = false;
	}
    
	public void run(){
	    this.last = null; //initially no blurb loaded
	    while (running){
		try{
		    this.doc = Jsoup.connect(this.url).get(); //gets the page
		} catch(Exception e) {
		    e.printStackTrace();
		}
		this.blurbs = this.doc.select("div[class=pb]"); //creates a list of blurbs
		this.latest = new Blurb(this.blurbs.first()); //gets first blurb
		this.blurbString = this.latest.toString(); //gets blurb in string form
		if (this.last == null || !(this.last.toString().equals(this.latest.toString()))){
		    //checks for updates to blurb
		    this.last = this.latest; //updates last if there is a new blurb
		    if (this.last.isComplete()){
			//if blurb is complete (has impact), proceed
			if (!(toEmail.equals("")))
			    mailBlurb(blurbString);
			//email blurb if we have an email address
			if (!(toText.equals("")))
			    textBlurb(blurbString);
			//texts blurb if we have a text email address
		    }
		    
		}
		try{
		    Thread.sleep(60000); //sleeps for a minute
		} catch (Exception e) {
		    e.printStackTrace();
		} 
	    }
	}
    
	public static void mailBlurb(String text){
	    SendEmail.send(fromEmail, toEmail, password, text);
	    //sends the email
	}

	public static void textBlurb(String text){
	    for (int i = 0; i < text.length(); i+= 159){
		if (i + 159 > text.length())
		    SendEmail.send(fromEmail, toText, password, text.substring(i, text.length()));
		//sends email as text message in 160-character blocks
		else
		    SendEmail.send(fromEmail, toText, password, text.substring(i, i + 159));
		//if we are at the end of the string, we send the rest
	    }
	}
    }


gui/GUIDriver.java
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.*;
import java.util.Scanner;
import java.lang.Thread;
import javax.swing.*;
import javax.swing.border.*;
import java.awt.*;
import java.awt.event.*;
class RotoScanGUI extends JPanel{

    Scanner scan = new Scanner(System.in);
    String input;
    String email = "";
    String text = "";
    String to = "";
    String password = ""; //initial strings that are initialized
    String NFLURL = "http://www.rotoworld.com/playernews/nfl/football/";
    String MLBURL = "http://www.rotoworld.com/playernews/mlb/baseball/";
    String NBAURL = "http://www.rotoworld.com/playernews/nba/basketball/";
    String NHLURL = "http://www.rotoworld.com/playernews/nhl/hockey/";
    String PGAURL = "http://www.rotoworld.com/playernews/gol/golf/";
    String NASCARURL = "http://www.rotoworld.com/playernews/nas/NASCAR/";
    String CFBURL = "http://www.rotoworld.com/playernews/cfb/nfl-draft"; //URLs

    Thread NFLThread;
    Thread MLBThread;
    Thread NBAThread;
    Thread NHLThread;
    Thread PGAThread;
    Thread NASCARThread;
    Thread CFBThread; //threads

    ScanRoto NFLRoto;
    ScanRoto MLBRoto;
    ScanRoto NBARoto;
    ScanRoto NHLRoto;
    ScanRoto PGARoto;
    ScanRoto NASCARRoto;
    ScanRoto CFBRoto; //ScanRoto objects

    private JCheckBox NFLBox;
    private JCheckBox MLBBox;
    private JCheckBox NBABox;
    private JCheckBox NHLBox;
    private JCheckBox PGABox;
    private JCheckBox NASCARBox;
    private JCheckBox CFBBox; //checkboxes for various sections of Rotoworld

    private JPanel left;
    private JPanel right; //left and right sides of the window

    private JCheckBox emailBox;
    private JLabel emailLabel;
    private JCheckBox textBox;
    private JLabel textLabel;
    private JTextField emailField;
    private JTextField textField;
    private JTextField yourEmailField;
    private JLabel yourEmailLabel;
    private JLabel passwordLabel;
    private JPasswordField passwordField;
    private JButton startStop; //right-side components

    void setup(){

	this.setLayout(new FlowLayout());
	left = new JPanel(new GridLayout(7, 1)); //left side is just checkboxes
	NFLBox = new JCheckBox("Football");
	left.add(NFLBox);
	MLBBox = new JCheckBox("Baseball");
	left.add(MLBBox);
	NBABox = new JCheckBox("Basketball");
	left.add(NBABox);
	NHLBox = new JCheckBox("Hockey");
	left.add(NHLBox);
	PGABox = new JCheckBox("Golf");
	left.add(PGABox);
	NASCARBox = new JCheckBox("NASCAR");
	left.add(NASCARBox);
	CFBBox = new JCheckBox("College Football");
	left.add(CFBBox); //creates checkboxes and adds them to left side

	right = new JPanel(new GridLayout(11,1,0,5));
	emailBox = new JCheckBox("Send email?");
	right.add(emailBox);
	textBox = new JCheckBox("Send text message?");
	right.add(textBox);
	emailLabel = new JLabel("Destination email address:");
	right.add(emailLabel);
	emailField = new JTextField();
	right.add(emailField);
	textLabel = new JLabel("Destination text message address:");
	right.add(textLabel);
	textField = new JTextField();
	right.add(textField);
	yourEmailLabel = new JLabel("Your email address:");
	right.add(yourEmailLabel);
	yourEmailField = new JTextField();
	right.add(yourEmailField);
	passwordLabel = new JLabel("Your email password:");
	right.add(passwordLabel);
	passwordField = new JPasswordField();
	right.add(passwordField);
	startStop = new JButton("Start!");
	right.add(startStop); //adds right-side components
	
	startStop.addActionListener(new ActionListener(){
		public void actionPerformed(ActionEvent e){
		    if (startStop.getText().equals("Start!")){
			startStop.setText("Stop!");
			if (emailBox.isSelected())
			    to = emailField.getText(); //gets destination address
			else
			    to = ""; //sets destination email
			email = yourEmailField.getText();
			password = passwordField.getText(); //gets password
			if (textBox.isSelected())
			    text = textField.getText();
			else
			    text = ""; //gets text message address
			if (NFLBox.isSelected()){
			    NFLRoto = new ScanRoto(NFLURL, email, to, password, text);
			    NFLThread = new Thread(NFLRoto, "NFL");
			    NFLThread.start();
			} //creates NFL thread

			if (MLBBox.isSelected()){
			    MLBRoto = new ScanRoto(MLBURL, email, to, password, text);
			    MLBThread = new Thread(MLBRoto, "MLB");
			    MLBThread.start();
			} //creates MLB thread
			    
			if (NBABox.isSelected()){
			    NBARoto = new ScanRoto(NBAURL, email, to, password, text);
			    NBAThread = new Thread(NBARoto, "NBA");
			    NBAThread.start();
			} //creates NBA thread

			if (NHLBox.isSelected()){
			    NHLRoto = new ScanRoto(NHLURL, email, to, password, text);
			    NHLThread = new Thread(NHLRoto, "NHL");
			    NHLThread.start();
			} //creates NHL thread

			if (PGABox.isSelected()){
			    PGARoto = new ScanRoto(PGAURL, email, to, password, text);
			    PGAThread = new Thread(PGARoto, "PGA");
			    PGAThread.start();
			} //creates PGA thread

			if (NASCARBox.isSelected()){
			    NASCARRoto = new ScanRoto(NASCARURL, email, to, password, text);
			    NASCARThread = new Thread(NASCARRoto, "NASCAR");
			    NASCARThread.start();
			} //creates NASCAR thread

			if (CFBBox.isSelected()){
			    CFBRoto = new ScanRoto(CFBURL, email, to, password, text);
			    CFBThread = new Thread(CFBRoto, "CFB");
			    CFBThread.start();
			} //creates college football thread
			NFLBox.setEnabled(false);
			MLBBox.setEnabled(false);
			NBABox.setEnabled(false);
			NHLBox.setEnabled(false);
			PGABox.setEnabled(false);
			NASCARBox.setEnabled(false);
			CFBBox.setEnabled(false);
			emailBox.setEnabled(false);
			textBox.setEnabled(false);
			//disables all checkboxes once the start button is clicked
			emailField.setEnabled(false);
			textField.setEnabled(false);
			yourEmailField.setEnabled(false);
			passwordField.setEnabled(false);
			//disables all text fields once start button is clicked
			
		    }

		    else{
			if (NFLBox.isSelected())
			    NFLRoto.stop();
			if (MLBBox.isSelected())
			    MLBRoto.stop();
			if (NBABox.isSelected())
			    NBARoto.stop();
			if (NHLBox.isSelected())
			    NHLRoto.stop();
			if (PGABox.isSelected())
			    PGARoto.stop();
			if (NASCARBox.isSelected())
			    NASCARRoto.stop();
			if (CFBBox.isSelected())
			    CFBRoto.stop(); //stops all running threads
			startStop.setText("Start!");
			NFLBox.setEnabled(true);
			MLBBox.setEnabled(true);
			NBABox.setEnabled(true);
			NHLBox.setEnabled(true);
			PGABox.setEnabled(true);
			NASCARBox.setEnabled(true);
			CFBBox.setEnabled(true);
			emailBox.setEnabled(true);
			textBox.setEnabled(true); //enables all checkboxes
			emailField.setEnabled(true);
			textField.setEnabled(true);
			yourEmailField.setEnabled(true);
			passwordField.setEnabled(true); //enables all text fields

			
		    }
		}
	    });
	
	this.add(left, "West");
	this.add(right, "East"); //adds the left and right sides
	
    }
    public RotoScanGUI(){
	setup(); //we create the panel
    }

}

public class GUIDriver extends CloseableFrame{
    public GUIDriver(){
	setTitle("RotoScan"); //sets window title
	setPreferredSize(new Dimension (500,375)); //sets initial window size
	Container contentPane = getContentPane();
	contentPane.add(new RotoScanGUI()); //adds components
	pack();
	setVisible(true); //draws frame
    }
    public static void main(String[] args){
	GUIDriver g = new GUIDriver(); //creates our frame
    }
}

class ScanRoto implements Runnable{
    private String url; //url for the thread
    private static String fromEmail; //source email address
    private static String toEmail; //destination email address
    private static String toText; //destination text messaging address
    private static String password; //GMail password 
    private Document doc; //document for the page
    private Blurb last; //the previous blurb
    private Blurb latest; //the latest blurb
    private Elements blurbs; //all of the blurbs on the page
    private String blurbString; //the string form of the most recent blurb
    private boolean running;
    public ScanRoto(String url, String fromEmail, String toEmail, String password, String text){
	this.url = url;
	this.fromEmail = fromEmail;
	this.toEmail = toEmail;
	this.password = password;
	this.toText = text; //stores local variables
	this.running = true; //sets thread state to running
    }

    public void stop(){
	running = false;
    }
    
    public void run(){
	this.last = null; //initially no blurb loaded
	while (running){
	    try{
		this.doc = Jsoup.connect(this.url).get(); //gets the page
	    } catch(Exception e) {
		e.printStackTrace();
	    }
	    this.blurbs = this.doc.select("div[class=pb]"); //creates a list of blurbs
	    this.latest = new Blurb(this.blurbs.first()); //gets first blurb
	    this.blurbString = this.latest.toString(); //gets blurb in string form
	    if (this.last == null || !(this.last.toString().equals(this.latest.toString()))){
		//checks for updates to blurb
		this.last = this.latest; //updates last if there is a new blurb
		if (this.last.isComplete()){
		    //if blurb is complete (has impact), proceed
		    if (!(toEmail.equals("")))
			mailBlurb(blurbString);
		    //email blurb if we have an email address
		    if (!(toText.equals("")))
			textBlurb(blurbString);
		    //texts blurb if we have a text email address
		}
		    
	    }
	    try{
		Thread.sleep(60000); //sleeps for a minute
	    } catch (Exception e) {
		e.printStackTrace();
	    } 
	}
    }
    
    public static void mailBlurb(String text){
	SendEmail.send(fromEmail, toEmail, password, text);
	//sends the email
    }

    public static void textBlurb(String text){
	for (int i = 0; i < text.length(); i+= 159){
	    if (i + 159 > text.length())
		SendEmail.send(fromEmail, toText, password, text.substring(i, text.length()));
	    //sends email as text message in 160-character blocks
	    else
		SendEmail.send(fromEmail, toText, password, text.substring(i, i + 159));
	    //if we are at the end of the string, we send the rest
	}
    }
}
